### ISSUE TITLE
[$100] Caps lock affects non-alphabetic keys

### ISSUE THREAD

**Issue by abridgewater** [robert-strandh/McCLIM#175]
Seems like it's behaving as Shift Lock, not Caps Lock, so on my (US QWERTY layout) keyboard, 5 becomes % and , becomes <.

**Comment by dkochmanski** [comment]
Some scraps from IRC:

```
< nyef`> The caps lock modifier needs to affect "A" but not "3". A shift lock modifier would need to affect both.
< nyef`> The implication that I'm seeing is that caps lock needs to apply only if the key name is #'ALPHA-CHAR-P, while shift-lock always applies.
< nyef`> ... Except that we then also need to consider the implications of the actual SHIFT keys.
< nyef`> ... And possibly *both* SHIFT keys (and there's an implicit assumption here that there are only *two* SHIFT keys).
< nyef`> ... Seems like there are (at least) three scenarios here. One is straight X11 core protocol, one is the X Keyboard Extension, and one is the X Input  Extension.
< nyef`> It would not surprise me in the least if CLX is limited to the X11 core protocol here.
```

**Comment by abridgewater** [comment]
Small amount of missing context for the above IRC scrap (the immediately preceding line):

```<nyef`> Oh, FFS. "3" on my keyboard is keycode 12.  Keysyms 0-3 are 51 35 51 35. "A" on my keyboard is keycode 38.  Keysyms 0-3 are 97 65 97 65.```

The point here being that they have the exact same pattern of keysyms (unshifted, shifted, unshifted, shifted), but require different handling with respect to shift lock and caps lock.

**Comment by abridgewater** [comment]
More from IRC:
```
<nyef`> Aha! From the CLX manual: "When the :lock modifier is on and is interpreted as :caps-lock, it is suggested that the :shift modifier first be applied to choose a keysym, but if that keysym is lowercase alphabetic, the corresponding uppercase keysym should be used instead." Section 15.4.1 Keycodes and Keysyms in my local copy.
<nyef`> Hrm. That doesn't seem quite right.
<nyef`> Shift + caps-lock on an alpha key needs to produce minuscules, but the logic described by the CLX manual would produce majuscules.
<nyef`> Still, this should be enough for someone to hack up Backends/CLX/keysyms.lisp / X-EVENT-TO-KEY-NAME-AND-MODIFIERS to do something more reasonably sane.
```
And that's that. I'm willing to test potential fixes once someone implements one.

**Comment by MatthewRock** [comment]
Hi! I have started working on this issue.

It is the first time I have encountered shift lock key, and I just want to make sure I got this right.

 I looked around the internet, and it looks like shift lock key is equivalent of keeping the shift down("sticky" shift). I have never used a shift lock (my keyboard don't have it anyway, but I suppose I could set other key to it), so I don't have much experience, but I would like to know if that's how it works:

(a) shift-lock OFF, caps-lock OFF, shift-key OFF: "normal"(lowercase) letters/characters
(b) shift-lock ON, caps-lock OFF, shift-key OFF: equivalent to pressing the shift
(c) shift-lock ON, caps-lock ON, shift-key OFF: Equivalent of pressing caps-lock and shift key at once: alphabetic keys are lowercase, other keys are equivalent of just pressing shift.
(d) shift-lock ON, caps-lock ON, shift-key ON: I'm not sure here. I guess this would mean it behaves just like caps-lock ON and any of these two shifts ON (e.g. (c))?
(e) shift-lock ON, caps-lock OFF, shift-key ON: Just like pressing single shift ((f), (b))?
(f) shift-lock OFF, caps-lock OFF, shift-key ON: Equivalent of pressing the shift ((b)).
(g) shift-lock OFF, caps-lock ON, shit-key OFF: Only alphabetic keys are upcase.
(h) shift-lock OFF, caps-lock ON, shift-key ON: alphabetic keys are downcase, other are upcase (just like (c))?

I am not really sure about (d), (e) and (h). I hope I got them right.

Unless these two keys can't be pressed together? I guess this is the current solution; lock-is-shift-lock-key-p (line 230, Backends/CLX/keysyms-common.lisp) returns T only if caps-lock is not already pressed, meaning it would not take these two pressed simultaneously into account.

Anyway, right now I am planning on fixing it like I said above: basically Shift-lock works just like an ordinary Shift key(but you don't have to hold anything) and can be pressed together with caps-lock. This would mean small changes in keysyms-common.lisp. Is it the right course of action?

**Comment by abridgewater** [comment]
I have a weak argument that the case of shift-lock ON and caps-lock ON will
never occur: What keyboard would have both?
I am not currently prepared to go to the trouble of parsing through the X
core protocol specification, the X Keyboard Extension specification, and
the X Input Extension specification to figure out if it is even *possible*
to have Lock-is-Caps-Lock and Lock-is-Shift-Lock at the same time, and if
so what the semantics are (you now know what the primary references for
this are if you didn't already, but feel free to introduce good secondary
references to this discussion).
With Caps-Lock and Shift, alphabetic keys are minuscules (lower-case),
other keys are shifted. With Shift-Lock and Shift, I have little idea. I
might argue that Shift-Lock + Shift behaves as Shift, and releasing Shift
also cancels Shift-Lock (justification: I've used typewriters with this
behaviour)!

**Comment by MatthewRock** [comment]
I've tested the behavior of shift-lock and caps lock on my PC, Gentoo Linux with X server. It looks like there's nothing wrong with the Shift Lock + Caps Lock at once, and the result was as I suspected - Shift Lock is equivalent of holding down the Shift key(but the Shift key won't turn it off - sorry!). This means that Any scenario with shift-lock should be treated as the user was holding the shift key itself, and the shift key does not matter at that point(since you can't "undo" the shift-lock with shift). I will implement this behavior.

**Comment by dkochmanski** [comment]
Fixed by #281.
